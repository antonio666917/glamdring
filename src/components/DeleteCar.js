import React from "react";

function DeleteCar({ id, children }) {
  const deleteCar = function () {
    return true;
  };

  return (
    <button
      type="button"
      onClick={() => {
        if (
          confirm("Are you sure you want to remove this car from your gallery?")
        ) {
          deleteCar().catch((err) => {
            alert(err.message);
          });
        }
      }}
    >
      {children}
    </button>
  );
}

export default DeleteCar;
