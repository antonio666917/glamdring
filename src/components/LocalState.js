import { createContext, useState, useContext } from "react";

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CheckoutStateProvider({ children }) {
  const [checkoutOpen, setCheckoutOpen] = useState(false);

  function toggleCheckout() {
    setCheckoutOpen(!checkoutOpen);
  }

  function closeCheckout() {
    setCheckoutOpen(false);
  }

  function openCheckout() {
    setCheckoutOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ checkoutOpen, toggleCheckout, openCheckout, closeCheckout }}
    >
      {children}
    </LocalStateProvider>
  );
}

function useCheckout() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CheckoutStateProvider, LocalStateContext, useCheckout };
