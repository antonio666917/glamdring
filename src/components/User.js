import gql from "graphql-tag";
import { useQuery } from "@apollo/client";

const CURRENT_USER_QUERY = gql`
  query {
    authenticatedUser {
      id
      email
      name
      permissions
      checkout {
        id
        quantity
        car {
          id
          price
          image {
            publicUrlTransformed
          }
          name
          description
        }
      }
    }
  }
`;

function useUser() {
  const { data, loading, error } = useQuery(CURRENT_USER_QUERY);
  if (data) {
    return data.authenticatedUser;
  }
}

export { CURRENT_USER_QUERY, useUser };
