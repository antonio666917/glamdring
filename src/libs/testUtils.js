import casual from "casual";

casual.seed(666);

const fakeCar = () => ({
  // __typename: 'Car',
  id: "honda-accord-ex-2020",
  mrsp: 29520,
  user: null,
  image: {
    publicUrlTransformed: "honda-accord-ex-2020.jpg",
  },
  name: "Honda Accord EX",
  description: "Honda Accord EX",
});

const fakeUser = (overrides) => ({
  // __typename: 'User',
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  ...overrides,
});

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export { LocalStorageMock, fakeCar, fakeUser };
