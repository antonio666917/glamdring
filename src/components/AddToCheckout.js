import React from "react";
import { CURRENT_USER_QUERY } from "./User";

const ADD_TO_CART_MUTATION = ``;

function AddToCheckout({ id }) {
  const loading = true;
  const addToCart = function () {
    return true;
  };
  return (
    <button type="button" disabled={loading} onClick={addToCart}>
      Add{loading && "ing!"} To Gallery
    </button>
  );
}

export default AddToCheckout;
export { ADD_TO_CART_MUTATION };
