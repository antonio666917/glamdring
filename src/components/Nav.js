import Link from "next/link";
import { useUser } from "./User";
import NavStyles from "../styles/NavStyles";
import { useCheckout } from "./LocalState";
import Signout from "./Signout";

function Nav() {
  const me = useUser();
  const { toggleCheckout } = useCheckout();
  return (
    <NavStyles data-testid="nav">
      <Link href="/used-cars">
        <a>Used Cars</a>
      </Link>
      {me && (
        <>
          <Link href="/track-your-order">
            <a>Track Your Order</a>
          </Link>
          <Link href="/me">
            <a>Account</a>
          </Link>
          <Signout />

          <button type="button" onClick={toggleCheckout}>
            Checkout
          </button>
        </>
      )}
      {!me && (
        <Link href="/signup">
          <a>Sign In</a>
        </Link>
      )}
    </NavStyles>
  );
}

export default Nav;
