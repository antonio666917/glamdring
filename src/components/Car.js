import React from "react";
import PropTypes from "prop-types";
import Link from "next/link";
import Title from "../styles/Title";
import CarStyles from "../styles/CarStyles";
import PriceTag from "../styles/PriceTag";
import formatMoney from "../libs/formatMoney";
import DeleteCar from "./DeleteCar";
import AddToCheckout from "./AddToCheckout";

export default function Car({ car }) {
  return (
    <CarStyles>
      {car.image && <img src={car.image.publicUrlTransformed} alt={car.name} />}

      <Title>
        <Link
          href={{
            pathname: "/car",
            query: { id: car.id },
          }}
        >
          <a>{car.name}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(car.price)}</PriceTag>
      <p>{car.description}</p>

      <div className="buttonList">
        <Link
          href={{
            pathname: "update",
            query: { id: item.id },
          }}
        >
          <a>Edit</a>
        </Link>
        <AddToCheckout id={car.id} />
        <DeleteCar id={car.id}>Remove from Checkout</DeleteCar>
      </div>
    </CarStyles>
  );
}

Car.propTypes = {
  Car: PropTypes.object.isRequired,
};
