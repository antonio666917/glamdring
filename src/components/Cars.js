import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import Car from "./Car";
import Pagination from "./Pagination";
import { perPage } from "../config";

// TODO:  orderBy: "createdAt_DESC"
const ALL_ITEMS_QUERY = ``;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${(props) => props.theme.maxWidth};
  margin: 0 auto;
`;

function Items({ page, count }) {
  const { data, error, loading } = [{}, {}, true];
  return (
    <Center>
      <Pagination page={page} />
      {(() => {
        if (loading) return <p>Loading...</p>;
        if (error) return <p>Error: {error.message}</p>;
        if (!data) return <p>No cars available at the moment</p>;
        return (
          <ItemsList>
            {data.allItems.map((item) => (
              <Item item={item} key={item.id} />
            ))}
          </ItemsList>
        );
      })()}
      <Pagination page={page} />
    </Center>
  );
}

Items.propTypes = {
  page: PropTypes.number,
};

export default Items;
export { ALL_ITEMS_QUERY };
