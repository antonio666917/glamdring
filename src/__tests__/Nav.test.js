import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser } from "../libs/testUtils";
import { CheckoutStateProvider } from "../components/LocalState";

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedUser: fakeUser() } },
  },
];

describe("<Nav/>", () => {
  it("Renders a shorter nav bar when users are signed out.", async () => {
    const { container } = render(
      <CheckoutStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CheckoutStateProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders a full nav bar when users are signed in.", async () => {
    const { container } = render(
      <CheckoutStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CheckoutStateProvider>
    );
    await screen.findByText("Account");
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent("Sign Out");
    expect(container).toHaveTextContent("My MFFFing Cart");
  });
});
