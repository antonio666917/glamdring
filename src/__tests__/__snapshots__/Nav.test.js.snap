// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> Renders a full nav bar when users are signed in. 1`] = `
<div>
  <ul
    class="NavStyles-sc-1gfalrs-0 cnYpHa"
    data-testid="nav"
  >
    <a
      href="/used-cars"
    >
      Used Cars
    </a>
    <a
      href="/track-your-order"
    >
      Track Your Order
    </a>
    <a
      href="/me"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign Out
    </button>
    <button
      type="button"
    >
      Checkout
    </button>
  </ul>
</div>
`;

exports[`<Nav/> Renders a shorter nav bar when users are signed out. 1`] = `
<div>
  <ul
    class="NavStyles-sc-1gfalrs-0 cnYpHa"
    data-testid="nav"
  >
    <a
      href="/used-cars"
    >
      Used Cars
    </a>
    <a
      href="/signup"
    >
      Sign In
    </a>
  </ul>
</div>
`;
